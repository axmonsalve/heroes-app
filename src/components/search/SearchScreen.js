import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {
  //manejando querystrings de busqueda
  const location = useLocation(); //Hook de react-router
  const { q = '' } = queryString.parse(location.search); //si es udefined lo asigno como string vacio

  const initialForm = {
    searchText: q,
  };
  const [formValues, handleInputChange] = useForm(initialForm);
  const { searchText } = formValues;

  //useMemo para no ejecutar cada vez que cambiar el state de la busqueda
  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    history.push(`?q=${searchText}`);
  };
  return (
    <div>
      <h1>searchScreen</h1>
      <hr />
      <div className="row">
        <div className="col-5 ">
          <h4>Search Form</h4>
          <form onSubmit={handleSearch}>
            <input
              type="text"
              autoComplete="off"
              className="form-control"
              name="searchText"
              value={searchText}
              onChange={handleInputChange}
              placeholder="Find your hero"
            />
            <button type="submit" className="btn mt-3 btn-primary col-12">
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {q === '' && <div className="alert alert-info">Search a Hero</div>}
          {q !== '' && heroesFiltered.length === 0 && (
            <div className="alert alert-danger">There is no Hero with {q}</div>
          )}

          {heroesFiltered.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </div>
  );
};
