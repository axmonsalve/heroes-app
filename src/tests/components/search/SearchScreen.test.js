import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe("Pruebas en SearchScrean", () => {
    const history = {
        push: jest.fn(),
    };
    test("Debe de mostrarse correctamente con valores por defecto", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe("Search a Hero");
    });

    test("Debe mostrar a batman y el input del valor de querystring", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find("input").prop("value")).toBe("batman");
        // expect(wrapper.find("HeroCard").exists()).toBe(true);
        expect(wrapper).toMatchSnapshot();
    });

    test("Debe mostrar un error si no encuentra el hero", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman123"]}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find("HeroCard").exists()).toBe(false);
        expect(wrapper.find(".alert-danger").text().trim()).toBe(
            "There is no Hero with batman123"
        );
    });
    test("Debe de llamar el push del history", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <Route
                    path="/search"
                    component={() => <SearchScreen history={history} />}
                />
            </MemoryRouter>
        );

        //simular el cambio en la caja de texto
        wrapper.find("input").simulate("change", {
            target: {
                name: "searchText",
                value: "batman",
            },
        });
        wrapper.find("form").prop("onSubmit")({
            preventDefault() {},
        });
        expect(history.push).toHaveBeenCalledWith("?q=batman");
    });
});
