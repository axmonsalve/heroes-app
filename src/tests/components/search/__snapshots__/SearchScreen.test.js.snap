// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScrean Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    searchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5 "
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn mt-3 btn-primary col-12"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en SearchScrean Debe mostrar a batman y el input del valor de querystring 1`] = `
<div>
  <h1>
    searchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5 "
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn mt-3 btn-primary col-12"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card mb-3 animate__animated animate__backInUp animate__faster"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img-top"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
